src/main/java/net/sf/mzmine/modules/peaklistmethods/io/siriusexport/MergeUtils.java

```diff
@@ -0,0 +1,1156 @@
+        // remember how many scans we had initially
+        int numberOfTotalScans = 0;
+        for (FragmentScan f : scans) numberOfTotalScans += f.ms2ScanNumbers.length;
+
+        final int numberOfScansRemovedDueToLowQuality = numberOfTotalScans - mzOrderedCopy.size();
+        final List<Integer> usedScanNumbers = new ArrayList<>();
+        // now, start from best scan and merge in both direction if cosine is large enough. For simplicity, we only
+        // compute cosine between the original scans, not the merged scan. This is slightly more efficient.
+        int numberOfMerges = 1;
+        DataPoint[] dataPoints = mzOrderedCopy.get(bestScan);
+        usedScanNumbers.add(scanNumbers.get(bestScan));
+        MergeDataPoint[] mergedSpectrum = new MergeDataPoint[dataPoints.length];
+        {
+            for (int k = 0; k < dataPoints.length; ++k) mergedSpectrum[k] = new MergeDataPoint(dataPoints[k]);
+        }
+
+        int i = bestScan - 1, j = bestScan + 1;
+        while (i > 0 || j < mzOrderedCopy.size()) {
+            if (i > 0) {
+                MergeDataPoint[] merged = merge(mzOrderedCopy.get(bestScan), mergedSpectrum, mzOrderedCopy.get(i), cosines[bestScan], cosines[i], cosineThreshold, parentPeak);
+                if (merged != null) {
+                    mergedSpectrum = merged;
+                    ++numberOfMerges;
+                    usedScanNumbers.add(scanNumbers.get(i));
+                }
+                --i;
+            }
+            if (j < mzOrderedCopy.size()) {
+                MergeDataPoint[] merged = merge(mzOrderedCopy.get(bestScan), mergedSpectrum, mzOrderedCopy.get(j), cosines[bestScan], cosines[j], cosineThreshold, parentPeak);
+                if (merged != null) {
+                    mergedSpectrum = merged;
+                    ++numberOfMerges;
+                    usedScanNumbers.add(scanNumbers.get(j));
+                }
+                ++j;
+            }
+        }
+
+        final int numberOfScansRemovedDueToLowCosine = mzOrderedCopy.size()-numberOfMerges;
+
+        // write a warning if too many scans are filtered out due to low cosine
+        if (mzOrderedCopy.size() >= 4 && ((double) numberOfMerges) / mzOrderedCopy.size() < 0.5) {
+            LoggerFactory.getLogger(MergeUtils.class).warn(name + ": Only " + numberOfMerges + " of " + mzOrderedCopy.size() + " spectra are merged. All other have a too low cosine similarity. Cosines between most intensive scan and other scans are: " + Arrays.toString(cosines));
+        }
+        MergedSpectrum M = new MergedSpectrum(mergedSpectrum, new RawDataFile[]{scans.get(0).origin}, scanNumbers.stream().mapToInt(x -> x).toArray(), polarityType, charge,numberOfScansRemovedDueToLowQuality, numberOfScansRemovedDueToLowCosine);
+        if (cutoffOutliers) M = cutOffOutliers(M);
+        if (!innerMerge) M = filterOutRarePeaks(M);
+        return M;
+    }
+
+
+    public List<FragmentScan> extractFragmentScansFor(PeakListRow row, Feature f) {
```

Comment:
There is a function ```f.getAllMS2FragmentScanNumbers()``` to extract all fragment scans - what else is this method doing? Perhaps some JavaDoc comments for the methods would be useful.

Reply:
f.getAllMS2FragmentScanNumbers() never did work for me. Maybe I have to reanalyze a project from raw data to see any effect?
Otherwise, this method will not only collect MSMS scans but also store the MS1 scans that precede the MSMS.

Reply:
It's a feature that was added recently, so if you're working with an older project file, the MS2 mapping won't be there. Reprocessing the raw data is one way to go, but Robin also made a separate module to just assign the MS2 scans to features: https://www.youtube.com/watch?v=EL5pmFvpTFE&feature=youtu.be

Reply:
If retrieving the MS1 scan before the MS2 is something that you need, perhaps we should add such function to the data model. What is the reason behind?

Reply:
Okay, I will check that.

So far there is only the possibility to get the "best" MSMS or all MSMS. What we need is a function that retrieves all "qualitative" MSMS spectra. The quality control could then be outsourced in an extra module.

The quality of a MSMS spectrum is affected by:
- intensity of the precursor ion in MS1
- how many high intensive peaks are within the isolation window if the precursor ion (chimeric spectra)
- total ion count of the MSMS spectrum
- how similar is the MSMS spectrum to other MSMS spectra of the same feature (cosine similarity)

Reply:
I like that approach! We can use the existing ```f.getAllMS2FragmentScanNumbers()``` function, and then have a separate module that can evaluate the quality of the MS2 spectra and for example provide a Comparator to sort them. Then the export module can choose to select e.g. the 5 highest quality spectra and export those.

Does the precursor ion intensity really matter? I think what matters is the TIC of the MSMS, but whether that TIC was produced from a high-intense precursor or not is irrelevant. Remember that the precursor scan might be quite distant (in terms of retention time) from the MSMS scan, or it might not even exist (targeted MS2) or there might be weird situations like MSe, all ion fragmentation, etc.